scx_rustland: enhancements

arighi opened the pull request:
Some fixes and enhancements to make scx_rustland more usable on a desktop / interactive system.

Decave commented on scheds/rust/scx_rustland/src/bpf/main.bpf.c:
```diff
@@ -282,6 +285,42 @@ static void dispatch_local(struct task_struct *p, u64 enq_flags)
+ */
+static s32 _get_task_cpu_slow(struct task_struct *p, s32 cpu)
+{
+	int i;
+
+	for (i = 0; i < num_possible_cpus; i++) {
+		cpu = (cpu + i) % num_possible_cpus;
+		if (bpf_cpumask_test_cpu(cpu, p->cpus_ptr))
+			return cpu;
+	}
+
+	return -1;
```
Something else you could do is call `bpf_cpumask_any_distribute(p->cpus_ptr)` to choose a random CPU from `p->cpus_ptr`? Note that it returns a cpu >= num_cpus if none of the bits are set, so you'd have to do something like:
```
cpu = bpf_cpumask_any_distribute(p->cpus_ptr);
return cpu < num_possible_cpus ? cpu : -1;
```

arighi replied:
Nice suggestion! Thanks! I'll apply it and run some tests. I think we should still maintain the logic to try the prev_cpu before picking a random one using `bpf_cpumask_any_distribute()`, that should reduce some unnecessary migrations, but I'll test/measure both and check if one shows some clear benefits vs the other.

Decave replied:
Yep, agreed. I was thinking that `bpf_cpumask_any_distribute()` could replace that for loop. But trying to choose prev_cpu first seems like a good idea.

Decave commented on scheds/rust/scx_rustland/src/bpf/main.bpf.c:
```diff
@@ -375,6 +411,14 @@ static bool is_task_cpu_available(struct task_struct *p, u64 enq_flags)
 
+	/*
+	 * Do not bypass the user-space scheduler if there are pending
+	 * activity, otherwise, we may risk to disrupt the scheduler's
```
optional suggestion:

s/we may risk to disrupt/we may risk disrupting

Decave reviewed (commented):
Overall looks good to me. Left one optional nit, and one suggestion for how we can make `_get_task_cpu_slow()` a lot faster.

arighi commented:
Moreover, I'm not yet fully convinced about the last 2 commits (change the default timeout_ms and slice_ns).

While they can definitely provide some benefits for interactive workloads I think there's also value to keep them as they are, because that can help to better measure future improvements of the scheduler.

Maybe I can add a command line option that would set them if a certain command line option is passed...

arighi commented:
Alright, I dropped some commits and updated this PR only with the "safe" fixes.

For the more experimental ones I'd like to do more testing (even if the other changes make the scheduler more responsive, they also make it more unpredictable sometimes, so I would like to investigate more on that).
